s3api/handlers_buckets.go

```diff
@@ -13,198 +14,134 @@ import (
+			if funcerr := f(); funcerr != nil {
+				log.Errorf("rollback task error: %s, continuing rollback", funcerr)
+			}
+		}
+	}
+}
+
 // BucketCreateHandler creates a new s3 bucket
 func (s *server) BucketCreateHandler(w http.ResponseWriter, r *http.Request) {
 	w = LogWriter{w}
 	vars := mux.Vars(r)
 	account := vars["account"]
 	s3Service, ok := s.s3Services[account]
 	if !ok {
-		log.Errorf("S3 service not found for account: %s", account)
-		w.WriteHeader(http.StatusNotFound)
+		msg := fmt.Sprintf("s3 service not found for account: %s", account)
+		handleError(w, apierror.New(apierror.ErrNotFound, msg, nil))
 		return
 	}
 
 	iamService, ok := s.iamServices[account]
 	if !ok {
-		log.Errorf("IAM service not found for account: %s", account)
-		w.WriteHeader(http.StatusNotFound)
+		msg := fmt.Sprintf("IAM service not found for account: %s", account)
+		handleError(w, apierror.New(apierror.ErrNotFound, msg, nil))
 		return
 	}
 
 	var req s3.CreateBucketInput
 	err := json.NewDecoder(r.Body).Decode(&req)
 	if err != nil {
-		log.Errorf("cannot decode body into create bucket input: %s", err.Error())
-		w.WriteHeader(http.StatusBadRequest)
+		msg := fmt.Sprintf("cannot decode body into create bucket input: %s", err)
+		handleError(w, apierror.New(apierror.ErrBadRequest, msg, err))
 		return
 	}
 
-	// Checks if a bucket exists in the account heading it.  This is a bit of a hack since in
-	// us-east-1 (only) bucket creation will succeed if the bucket already exists in your
-	// account.  In all other regions, the API will return s3.ErrCodeBucketAlreadyOwnedByYou 🤷‍♂️
-	_, err = s3Service.Service.HeadBucketWithContext(r.Context(), &s3.HeadBucketInput{
-		Bucket: req.Bucket,
-	})
-	if err == nil {
-		msg := fmt.Sprintf("%s: Bucket exists and is owned by you.", s3.ErrCodeBucketAlreadyOwnedByYou)
-		w.WriteHeader(http.StatusConflict)
-		w.Write([]byte(msg))
-		return
-	}
+	// setup rollback function list and defer recovery and execution
+	var rollBackTasks []func() error
+	defer func() {
+		if err := recover(); err != nil {
+			log.Errorf("recovering from panic: %s", err)
+			executeRollBack(&rollBackTasks)
+		}
+	}()
 
-	log.Infof("creating bucket: %s", *req.Bucket)
-	bucketOutput, err := s3Service.Service.CreateBucketWithContext(r.Context(), &req)
+	bucketOutput, err := s3Service.CreateBucket(r.Context(), &req)
 	if err != nil {
-		if aerr, ok := err.(awserr.Error); ok {
-			switch aerr.Code() {
-			case s3.ErrCodeBucketAlreadyExists:
-				msg := fmt.Sprintf("%s: %s", s3.ErrCodeBucketAlreadyExists, aerr.Error())
-				w.WriteHeader(http.StatusConflict)
-				w.Write([]byte(msg))
-			case s3.ErrCodeBucketAlreadyOwnedByYou:
-				msg := fmt.Sprintf("%s: %s", s3.ErrCodeBucketAlreadyOwnedByYou, aerr.Error())
-				w.WriteHeader(http.StatusConflict)
-				w.Write([]byte(msg))
-			default:
-				w.WriteHeader(http.StatusBadRequest)
-				w.Write([]byte(err.Error()))
+		handleError(w, err)
+		msg := fmt.Sprintf("failed to create bucket: %s", err)
+		panic(msg)
```

Comment:
I read somewhere that using `panic` is bad, but I assume here it's OK because we're using the `RecoveryHandler`?

Reply:
i think its okay, but i can do some research... lets move forward with this but we'll revisit before deploying this GA